**Hyper parameter tuning**

- The parameters or arguments of model can change by user to get the good model performance

- Grid search

- Random search

- Assume that in Decsision tree we have following hyper parameter

   - criterion : {"gini", "entropy", "log_loss"}

   - splitter : {"best", "random"}

   - max depth : {1,2,3,4,5,6}

- By default we have criterion 'gini' is there

- But the model might be performed when criterion is 'entropy' also

- we need to test the combinations by providing different values

- To try the combinations we have two methods

   - Grid search:

      - it is a step by step by consider the all combinations

      - in criterion: 3,splitter: 2, max depth=6

      - The total combinations are : 3x2x6=36

   - Random search:

      - It will not do all 36 possibile combiantions

      - It will select randomly some combinations

- Generally we do Grid search

- By doing this we have term called Cross validation

**Cross validation:**

- Cross validation: CV

- CV=4 means we devide data randomly into 4 parts

- Out of these 4 parts every time 3 parts consider as train data, 1 parts consider as test data

- P1 P2 P3 P4 are 4 parts of data

- Combination1: P1 + P2 + P3 are train data and P4 as test data

   - For Combination-1 will try to imporove model performance by doing Hyper parameter tuning

   - Assume there 36 comninations are available in DT

   - For the Combination1 will do all 36

   - so we have 36 hyper parameter comninations for each combination will have one accuracy

   - so we have total 36 accuracy values

- Combination2: P1 + P2 + P4 are train data and P3 as test data

   - For Combination-2 will try to imporove model performance by doing Hyper parameter tuning

   - Assume there 36 comninations are available in DT

   - For the Combination-2 will do all 36

- Combination3: P1 + P3 + P4 are train data and P2 as test data

   - For Combination-3 will try to imporove model performance by doing Hyper parameter tuning

   - Assume there 36 comninations are available in DT

   - For the Combination-3 will do all 36

- Combination-4: P2 + P3 + P4 are train data and P1 as test data

   - For Combination-4 will try to imporove model performance by doing Hyper parameter tuning

   - Assume there 36 comninations are available in DT

   - For the Combination-4 will do all 36

- Every combination gives 36 accurcy values corresponding each hyper parameter

- will calculate all the average accuracy of combinations individual accuracy w.r.t to each hyper parameter

- so that will get 36 average accuracy values w.r.t each hyper parameter

- Out of these 36 the maximum one is our best accuracy and corresponding hyper parameters are best parameters

In [2]:
import pandas as pd
indexes=[f"h_{i}" for i in range(1,37)]
col1=[f"combination_1_acc_{i}" for i in range(1,37)]
col2=[f"combination_2_acc_{i}" for i in range(1,37)]
col3=[f"combination_3_acc_{i}" for i in range(1,37)]
col4=[f"combination_4_acc_{i}" for i in range(1,37)]
col5=[f"avg_acc_{i}" for i in range(1,37)]
cols=['C1','C2','C3','C4','Avg acc']
pd.DataFrame(zip(col1,col2,col3,col4,col5),index=indexes,
            columns=cols)

C1                    C2                    C3  \
h_1    combination_1_acc_1   combination_2_acc_1   combination_3_acc_1   
h_2    combination_1_acc_2   combination_2_acc_2   combination_3_acc_2   
h_3    combination_1_acc_3   combination_2_acc_3   combination_3_acc_3   
h_4    combination_1_acc_4   combination_2_acc_4   combination_3_acc_4   
h_5    combination_1_acc_5   combination_2_acc_5   combination_3_acc_5   
h_6    combination_1_acc_6   combination_2_acc_6   combination_3_acc_6   
h_7    combination_1_acc_7   combination_2_acc_7   combination_3_acc_7   
h_8    combination_1_acc_8   combination_2_acc_8   combination_3_acc_8   
h_9    combination_1_acc_9   combination_2_acc_9   combination_3_acc_9   
h_10  combination_1_acc_10  combination_2_acc_10  combination_3_acc_10   
h_11  combination_1_acc_11  combination_2_acc_11  combination_3_acc_11   
h_12  combination_1_acc_12  combination_2_acc_12  combination_3_acc_12   
h_13  combination_1_acc_13  combination_2_acc_13  combination_3_acc_13   
h_14  combination_1_acc_14  combination_2_acc_14  combination_3_acc_14   
h_15  combination_1_acc_15  combination_2_acc_15  combination_3_acc_15   
h_16  combination_1_acc_16  combination_2_acc_16  combination_3_acc_16   
h_17  combination_1_acc_17  combination_2_acc_17  combination_3_acc_17   
h_18  combination_1_acc_18  combination_2_acc_18  combination_3_acc_18   
h_19  combination_1_acc_19  combination_2_acc_19  combination_3_acc_19   
h_20  combination_1_acc_20  combination_2_acc_20  combination_3_acc_20   
h_21  combination_1_acc_21  combination_2_acc_21  combination_3_acc_21   
h_22  combination_1_acc_22  combination_2_acc_22  combination_3_acc_22   
h_23  combination_1_acc_23  combination_2_acc_23  combination_3_acc_23   
h_24  combination_1_acc_24  combination_2_acc_24  combination_3_acc_24   
h_25  combination_1_acc_25  combination_2_acc_25  combination_3_acc_25   
h_26  combination_1_acc_26  combination_2_acc_26  combination_3_acc_26   
h_27  combination_1_acc_27  combination_2_acc_27  combination_3_acc_27   
h_28  combination_1_acc_28  combination_2_acc_28  combination_3_acc_28   
h_29  combination_1_acc_29  combination_2_acc_29  combination_3_acc_29   
h_30  combination_1_acc_30  combination_2_acc_30  combination_3_acc_30   
h_31  combination_1_acc_31  combination_2_acc_31  combination_3_acc_31   
h_32  combination_1_acc_32  combination_2_acc_32  combination_3_acc_32   
h_33  combination_1_acc_33  combination_2_acc_33  combination_3_acc_33   
h_34  combination_1_acc_34  combination_2_acc_34  combination_3_acc_34   
h_35  combination_1_acc_35  combination_2_acc_35  combination_3_acc_35   
h_36  combination_1_acc_36  combination_2_acc_36  combination_3_acc_36   

                        C4     Avg acc  
h_1    combination_4_acc_1   avg_acc_1  
h_2    combination_4_acc_2   avg_acc_2  
h_3    combination_4_acc_3   avg_acc_3  
h_4    combination_4_acc_4   avg_acc_4  
h_5    combination_4_acc_5   avg_acc_5  
h_6    combination_4_acc_6   avg_acc_6  
h_7    combination_4_acc_7   avg_acc_7  
h_8    combination_4_acc_8   avg_acc_8  
h_9    combination_4_acc_9   avg_acc_9  
h_10  combination_4_acc_10  avg_acc_10  
h_11  combination_4_acc_11  avg_acc_11  
h_12  combination_4_acc_12  avg_acc_12  
h_13  combination_4_acc_13  avg_acc_13  
h_14  combination_4_acc_14  avg_acc_14  
h_15  combination_4_acc_15  avg_acc_15  
h_16  combination_4_acc_16  avg_acc_16  
h_17  combination_4_acc_17  avg_acc_17  
h_18  combination_4_acc_18  avg_acc_18  
h_19  combination_4_acc_19  avg_acc_19  
h_20  combination_4_acc_20  avg_acc_20  
h_21  combination_4_acc_21  avg_acc_21  
h_22  combination_4_acc_22  avg_acc_22  
h_23  combination_4_acc_23  avg_acc_23  
h_24  combination_4_acc_24  avg_acc_24  
h_25  combination_4_acc_25  avg_acc_25  
h_26  combination_4_acc_26  avg_acc_26  
h_27  combination_4_acc_27  avg_acc_27  
h_28  combination_4_acc_28  avg_acc_28  
h_29  combination_4_acc_29  avg_acc_29  
h_30  combination_4_acc_30  avg_acc_30  
h_31  combinat

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme(color_codes=True)
pd.set_option('display.max_columns', None)
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,\
                            precision_score,\
                            recall_score,\
                            f1_score,\
                            classification_report,\
                            roc_auc_score,roc_curve,auc
from sklearn.metrics import confusion_matrix,ConfusionMatrixDisplay

In [4]:
data=pd.read_csv(r'C:\Users\sravani9\OneDrive\Documents\Naresh IT\Python sessions\Data_files\preprocessd_data.csv')
data    

gender       age  no_of_days_subscribed  multi_screen  mail_subscribed  \
0          0 -0.264643              -0.946659             0                0   
1          0  0.029654               1.235009             0                0   
2          0  2.580235               0.658246             0                0   
3          0 -1.441834               0.783629             0                1   
4          0  0.127754               2.288228             0                0   
...      ...       ...                    ...           ...              ...   
3401       1 -0.805656               0.479710             0                0   
3402       0  0.477826               0.674425             0                0   
3403       0 -0.550521               0.426099             0                1   
3404       0  0.004651              -0.444637             0                0   
3405       0 -0.057028               0.317748             0                0   

      weekly_mins_watched  minimum_daily_mins  weekly_max_night_mins  \
0               -1.508785            0.722258              -0.944404   
1                0.304821           -0.901748              -0.688523   
2               -2.266627            0.613991              -0.483818   
3                0.638123           -0.252146               0.079121   
4               -0.333851            0.253101              -0.893228   
...                   ...                 ...                    ...   
3401             0.887802            0.552184               0.865765   
3402             2.038061           -0.931071               0.567257   
3403            -0.410669            0.344847              -0.729485   
3404             1.359751           -0.847121              -0.715010   
3405             2.844852            0.098115              -0.550251   

      videos_watched  customer_support_calls  churn  
0          -1.403000               -0.415473      0  
1          -0.599124                0.347522      0  
2          -1.403000                2.636507      1  
3          -0.197185                1.110517      0  
4           1.008630               -0.415473      0  
...              ...                     ...    ...  
3401       -0.495634               -0.415473      1  
3402        2.100185                0.235227      1  
3403       -0.403904                1.873512      1  
3404       -0.197185               -1.178468      1  
3405        0.606692               -1.178468      1  

[3406 rows x 11 columns]

**divide data into input and output**

In [5]:
# Here the target column is Customer Status
X = data.drop('churn', axis=1)
y = data['churn']

**test and split**

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2,random_state=0)
print(X_train.shape),
print(X_test.shape),
print(y_train.shape),
print(y_test.shape)

(2724, 10)
(682, 10)
(2724,)
(682,)


**Read the base model**

In [7]:
from sklearn.model_selection import GridSearchCV,cross_val_score
from sklearn.tree import DecisionTreeClassifier
grid_tree=DecisionTreeClassifier()
grid_tree

DecisionTreeClassifier()

**Create a parameter file**

In [8]:
grid_tree.get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'random_state': None,
 'splitter': 'best'}

In [9]:
param_grid={
    "criterion":['gini','entropy'],
    'max_depth':[3, 4, 5, 6, 7, 8],
    'min_samples_split':[2, 3, 4],
    'min_samples_leaf':[1, 2, 3, 4],
    'random_state':[0, 42]
}

In [10]:
grid_search = GridSearchCV(grid_tree,
                           param_grid,
                           scoring='accuracy',
                           cv=5,
                           verbose=True)

In [11]:
import time
import warnings
warnings.filterwarnings("ignore")
start=time.time()
grid_search.fit(X_train,y_train)  
# Generally we will pass the entire data
# X_train= input data
# y_train= output data
end=time.time()
print("Total time taken is:",(end-start))

Fitting 5 folds for each of 288 candidates, totalling 1440 fits
Total time taken is: 20.17229127883911


In [11]:
dir(grid_search)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__sklearn_clone__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_build_request_for_signature',
 '_check_feature_names',
 '_check_n_features',
 '_check_refit_for_multimetric',
 '_doc_link_module',
 '_doc_link_template',
 '_doc_link_url_param_generator',
 '_estimator_type',
 '_format_results',
 '_get_default_requests',
 '_get_doc_link',
 '_get_metadata_request',
 '_get_param_names',
 '_get_routed_params_for_fit',
 '_get_scorers',
 '_get_tags',
 '_more_tags',
 '_parameter_constraints',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_required_parameters',
 '_run

In [17]:
grid_search.best_estimator_

DecisionTreeClassifier(criterion='entropy', max_depth=8, min_samples_leaf=4,
                       random_state=0)

**Estimator means model**

- after developed model 1440 times

  - 288 parameters combinations with CV of 5

  - each parameter combination trained 5 tyms

  - each paramete combination will get 5 accuracy values

  - each parameter combination will get one avg accuracy

  - like that we will get 288 avg accuracy values

  - in that one parameter combination has heighest avg accuracy

  - that is best parameter

  - the model is using best parameters called as bets estimator

In [12]:
grid_search.cv_results_.keys()

dict_keys(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time', 'param_criterion', 'param_max_depth', 'param_min_samples_leaf', 'param_min_samples_split', 'param_random_state', 'params', 'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score', 'mean_test_score', 'std_test_score', 'rank_test_score'])

In [13]:
len(grid_search.cv_results_['param_criterion'].data)

288

In [14]:
grid_search.cv_results_['params']

[{'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 2,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 2,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 3,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 3,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 4,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 1,
  'min_samples_split': 4,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_split': 2,
  'random_state': 0},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_split': 2,
  'random_state': 42},
 {'criterion': 'gini',
  'max_depth': 3,
  'min_samples_leaf': 2,
  'min_samples_spl

In [15]:
grid_search.cv_results_['split0_test_score']

array([0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82752294,
       0.82752294, 0.82752294, 0.82752294, 0.82752294, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82018349, 0.82018349, 0.82018349,
       0.82018349, 0.82018349, 0.82201835, 0.82201835, 0.82201835,
       0.82201835, 0.82201835, 0.82201835, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84587156, 0.84587156, 0.84587156, 0.84587156,
       0.84587156, 0.84770642, 0.84770642, 0.84770642, 0.84770642,
       0.84770642, 0.84770642, 0.83853211, 0.84220183, 0.83853

In [18]:
param1_acc1=grid_search.cv_results_['split0_test_score'][0]
param1_acc2=grid_search.cv_results_['split1_test_score'][0]
param1_acc3=grid_search.cv_results_['split2_test_score'][0]
param1_acc4=grid_search.cv_results_['split3_test_score'][0]
param1_acc5=grid_search.cv_results_['split4_test_score'][0]
print(param1_acc1,param1_acc2,param1_acc3,param1_acc4,param1_acc5)
param1_avg_acc=np.mean([param1_acc1,param1_acc2,param1_acc3,param1_acc4,param1_acc5])
param1_avg_acc
grid_search.cv_results_['mean_test_score']

0.8275229357798165 0.8293577981651377 0.8146788990825689 0.8256880733944955 0.8106617647058824


array([0.82158189, 0.82158189, 0.82158189, 0.82158189, 0.82158189,
       0.82158189, 0.82194887, 0.82194887, 0.82194887, 0.82194887,
       0.82194887, 0.82194887, 0.82194887, 0.82194887, 0.82194887,
       0.82194887, 0.82194887, 0.82194887, 0.82194887, 0.82194887,
       0.82194887, 0.82194887, 0.82194887, 0.82194887, 0.81827712,
       0.81754317, 0.81827712, 0.81754317, 0.81827712, 0.81754317,
       0.81937938, 0.81791015, 0.81937938, 0.81791015, 0.81937938,
       0.81791015, 0.81901241, 0.81827847, 0.81901241, 0.81827847,
       0.81901241, 0.81827847, 0.81974636, 0.81901241, 0.81974636,
       0.81901241, 0.81974636, 0.81901241, 0.84544253, 0.84544253,
       0.84544253, 0.84544253, 0.84544253, 0.84544253, 0.84434161,
       0.84581085, 0.84434161, 0.84581085, 0.84434161, 0.84581085,
       0.84581017, 0.84544253, 0.84581017, 0.84544253, 0.84581017,
       0.84544253, 0.84617647, 0.84617647, 0.84617647, 0.84617647,
       0.84617647, 0.84617647, 0.8417782 , 0.8421445 , 0.84177